### DISCUSSION TITLE
a difference in the output of namespace moby and k8s.io.

### DISCUSSION THREAD

**Discussion by Sungjun-Ohh** [containerd/nerdctl#1412] [Q&A] [answered]
### Description

There is a difference in the output of namespace moby and k8s.io.
Case1 is a Kubernetes service whose runtime is set to containerd.
Case2 is a Kubernetes service whose runtime is set to dockerd.

### Steps to reproduce the issue

* case1.  namespace=k8s.io
-------------------------------------
[user@hostname sungjun]$ sudo ./nerdctl --namespace=k8s.io ps -a
CONTAINER ID    IMAGE                                                                                                                 COMMAND                   CREATED         STATUS     PORTS    NAMES
01c1c77e8eaa    sha256:0184c1613d92931126feb4c548e5da11015513b9e4c104e7305ee8b53b50a9da                                               "/pause"                  5 months ago    Up                  k8s://kube-system/nodelocaldns-62cd5
05b05ddf11f6    k8s.gcr.io/pause:3.3                                                                                                  "/pause"                  8 days ago      Up                  k8s://kube-system/kube-controller-manager-hostname
0d64f2042014    reg.navercorp.com/base/alpine3/python@sha256:4c86d88c6bbad6cd9fd9ef57175121afd2dcff222e85d334b1ac1585149a3a32         "/bin/bash -c while …"    4 months ago    Up                  k8s://n3r-system/node-exporter-pqhbc/textcollector
1a39ef529831    k8s.gcr.io/pause:3.3                                                                                                  "/pause"                  5 months ago    Up                  k8s://kube-system/kube-apiserver-hostname
1ae0286c6ece    k8s.gcr.io/pause:3.3                                                                                                  "/pause"                  2 months ago    Up                  k8s://kube-system/cilium-rl2rw
1c21f2fde8e1    k8s.gcr.io/pause:3.3                                                                                                  "/pause"                  11 days ago     Up                  k8s://kube-system/dns-autoscaler-f8b6f8d5-2lzlj
2425c708d9a1    sha256:884d49d6d8c9f40672d20c78e300ffee238d01c1ccb2c132937125d97a596fd7                                               "kube-scheduler --au…"    5 months ago    Up                  k8s://kube-system/kube-scheduler-hostname/kube-scheduler
243800f1f0dd    reg.navercorp.com/ncc-public/cilium/cilium:v1.11.6-n3r-v3                                                             "cilium-agent --kvst…"    5 weeks ago     Created             k8s://kube-system/cilium-rl2rw/cilium-agent
37c2237e54ff    k8s.gcr.io/kube-apiserver:v1.23.5                                                                                     "kube-apiserver --ad…"    5 months ago    Up                  k8s://kube-system/kube-apiserver-hostname/kube-apiserver
37dc6ec9ff79    reg.navercorp.com/n3r-public/fluent-bit:v1.9.2-n2                                                                     "/fluent-bit/bin/flu…"    2 months ago    Up                  k8s://n3r-system/fluent-bit-j649j/fluent-bit
3a1099df4099    reg.navercorp.com/n3r-public/node-problem-detector:v0.8.10-n2-rc2                                                     "/node-problem-detec…"    8 weeks ago     Up                  k8s://n3r-system/node-problem-detector-2xcqd/node-problem-detector
4f9cd96989dd    reg.navercorp.com/ncc-public/node-exporter@sha256:49d3f4fc2e25d099d34fb157381426757f75ad7642082f0e50112ba4a684036e    "/bin/node_exporter …"    4 months ago    Up                  k8s://n3r-system/node-exporter-pqhbc/node-exporter
51bf9437b61e    k8s.gcr.io/pause:3.3                                                                                                  "/pause"                  3 months ago    Up                  k8s://n3r-system/duct-tape-6ftc5
55fd80ff5f0c    sha256:0184c1613d92931126feb4c548e5da11015513b9e4c104e7305ee8b53b50a9da                                               "/pause"                  5 months ago    Up                  k8s://kube-system/kube-scheduler-hostname
6007a2b0530a    k8s.gcr.io/pause:3.3                                                                                                  "/pause"                  2 months ago    Up                  k8s://n3r-system/fluent-bit-j649j
6168683e1d65    reg.navercorp.com/base/alpine:3.15                                                                                    "sh -c sleep 15"          3 months ago    Created             k8s://n3r-system/duct-tape-6ftc5/delay-start
62382dfbd356    reg.navercorp.com/n3r-public/node_auto_repair:v0.0.2                                                                  "/root/bin/docker-en…"    2 weeks ago     Up                  k8s://n3r-system/node-auto-repair-4b92b/node-auto-repair
672fa5501e2c    reg.navercorp.com/ncc-public/cilium/cilium:v1.11.6-n3r-v3                                                             "cilium-agent --kvst…"    4 weeks ago     Up                  k8s://kube-system/cilium-rl2rw/cilium-agent
6b17889f1ec8    k8s.gcr.io/dns/k8s-dns-node-cache@sha256:3c465863216783b5f0e7c65fa7d8420cbd4eb4b63383a10d5d3c965df623a820             "/node-cache -locali…"    5 months ago    Up                  k8s://kube-system/nodelocaldns-62cd5/node-cache
860dcab2098d    k8s.gcr.io/pause:3.3                                                                                                  "/pause"                  8 weeks ago     Up                  k8s://n3r-system/node-problem-detector-2xcqd
8f805852a134    k8s.gcr.io/cpa/cluster-proportional-autoscaler-amd64:1.8.3                                                            "/cluster-proportion…"    11 days ago     Up                  k8s://kube-system/dns-autoscaler-f8b6f8d5-2lzlj/autoscaler
a4dec45dc206    k8s.gcr.io/kube-controller-manager:v1.23.5                                                                            "kube-controller-man…"    8 days ago      Up                  k8s://kube-system/kube-controller-manager-hostname/kube-controller-manager
ab31a63a6156    reg.navercorp.com/n3r-public/duct-tape:v0.0.1-2                                                                       "duct-tape"               3 months ago    Up                  k8s://n3r-system/duct-tape-6ftc5/n3r-nc
acab4570a4ee    reg.navercorp.com/ncc-public/cilium/cilium:v1.11.6-n3r-v3                                                             "/init-container.sh"      2 months ago    Created             k8s://kube-system/cilium-rl2rw/clean-cilium-state
bf38f4ffb1ed    k8s.gcr.io/pause:3.3                                                                                                  "/pause"                  2 weeks ago     Up                  k8s://n3r-system/node-auto-repair-4b92b
d9874a6967b4    reg.navercorp.com/ncc-public/healoxy:v0.2.2-0-gaa39976                                                                "/usr/bin/healoxy --…"    5 months ago    Up                  k8s://kube-system/kube-apiserver-hostname/healoxy
f0b62687098b    k8s.gcr.io/pause:3.3                                                                                                  "/pause"                  4 months ago    Up                  k8s://n3r-system/node-exporter-pqhbc
f0f59722cc91    reg.navercorp.com/ncc-public/coredns/coredns:v1.7.0-n3r-a9f42aaf2                                                     "/coredns -conf /etc…"    12 days ago     Up                  k8s://kube-system/coredns-f47f5695d-mds4l/coredns
fa9eaa71abb5    k8s.gcr.io/pause:3.3                                                                                                  "/pause"                  12 days ago     Up                  k8s://kube-system/coredns-f47f5695d-mds4l

* case2.  namespace=moby 
-------------------------------------
[user@hostname nerdbin]$ sudo ./nerdctl --namespace=moby ps -a
CONTAINER ID    IMAGE    COMMAND                   CREATED         STATUS    PORTS    NAMES
01b130f2afc7             "/pause"                  8 months ago    Up                 
01bf2a9595c0             "/pause"                  8 months ago    Up                 
0735bbc3013d             "/cluster-proportion…"    8 months ago    Up                 
07894d73392d             "start_runit"             8 months ago    Up                 
114a04f20228             "sleep 99999999999d"      8 months ago    Up                 
182e33672b4b             "/manager --port=844…"    7 weeks ago     Up                 
25b8443b2f19             "/pause"                  8 months ago    Up                 
2616e27d4f7b             "/pause"                  3 months ago    Up                 
29a477ae103b             "/pause"                  8 months ago    Up                 
2a160a67e25c             "tail -f /dev/null"       3 months ago    Up                 
35d29af5341d             "kube-controller-man…"    16 hours ago    Up                 
4258cf16e15e             "/pause"                  8 months ago    Up                 
448a09853ba8             "/pause"                  8 months ago    Up                 
68b4a2d20459             "/pause"                  7 weeks ago     Up                 
6cf95fd3985a             "/pause"                  8 months ago    Up                 
6dd934f1bed1             "sleep 99999999999d"      8 months ago    Up                 
783ecb3b58a2             "tail -f /dev/null"       7 weeks ago     Up                 
78b20b06741e             "/coredns -conf /etc…"    8 months ago    Up                 
7dd851da5f29             "kube-apiserver --ad…"    3 weeks ago     Up                 
80863e5e1a9c             "/docker-entrypoint.…"    6 months ago    Up                 
87f9426cd735             "/pause"                  8 months ago    Up                 
8853c667922c             "/pause"                  7 weeks ago     Up                 
8d423ede2a9e             "sleep 99999999999d"      6 months ago    Up                 
94484c85f9ba             "/usr/local/bin/entr…"    3 months ago    Up                 
a5f1b7d2b4c9             "/usr/local/bin/kube…"    8 months ago    Up                 
a6fec5cecb6c             "kube-scheduler --au…"    16 hours ago    Up                 
aec899774c7c             "/pause"                  8 months ago    Up                 
c7f2fe6b0b2d             "/pause"                  7 weeks ago     Up                 
ccb0ddb553b8             "/docker-entrypoint.…"    8 months ago    Up                 
d7acd3b4f384             "/pause"                  8 months ago    Up                 
dc666ea8c15a             "/docker-entrypoint.…"    8 months ago    Up                 
e55070cf59fc             "/usr/local/bin/etcd"     8 months ago    Up                 
ea774cc3bbbe             "/pause"                  3 months ago    Up                 
edf2e678ef16             "sleep 9999999999999…"    8 months ago    Up                 
f78cbbb4a890             "/node-cache -locali…"    8 months ago    Up                 
f8ddd3f9d723             "/docker-entrypoint.…"    8 months ago    Up                 
f91600939130             "/pause"                  6 months ago    Up       

### Describe the results you received and expected

I expected the same output to come out even if the namespace is different. If the namespace is k8s.io, the output is printed as expected. However, I wanted the image name (or image digest) to be printed when the namespace is moby.

Also, in case 1, the form is not unified as shown below. 
a) imagename, b) imagedigest, c) imagename & digest

### What version of nerdctl are you using?

In both cases, I went with nerdctl version 0.23.0.

### Host information

* case1.  namespace=k8s.io
--------------
Client:
 Namespace:     default
 Debug Mode:    false

Server:
 Server Version: 1.4.9
 Storage Driver: overlayfs
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Log: fluentd journald json-file
  Storage: native overlayfs
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 5.15.27
 Operating System: CentOS Linux 7 (Core)
 OSType: linux
 Architecture: x86_64
 CPUs: 40
 Total Memory: 62.4GiB
 Name: (secret)
 ID: 98bc6381-41e3-4e32-b374-7e02f9c5c9a7

* case2.  namespace=moby 
------------------
Client:
 Namespace:     default
 Debug Mode:    false

Server:
 Server Version: 1.4.12
 Storage Driver: overlayfs
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Log: fluentd journald json-file
  Storage: native overlayfs
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 4.18.0-193.el8.x86_64
 Operating System: CentOS Linux 8
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 7.617GiB
 Name: (secret)
 ID: 9e874012-62ac-4d8c-a77e-fbc14a8f98d2

**Comment by AkihiroSuda** [comment, marked as answer]
> I wanted the image name (or image digest) to be printed when the namespace is moby.

Image information cannot be shown with Docker 20.10, as Docker 20.10 does not use containerd's image store.

Docker 23.XX will use containerd's image store, so the image names will be inspectable via nerdctl too.
https://github.com/moby/moby/pulls?q=is%3Apr+is%3Aopen+label%3Acontainerd-integration
